backend/src/services/portfolioService.ts

```
@@ -0,0 +1,193 @@
+import bcrypt from 'bcrypt';
+import cookieParser from 'cookie-parser';
+import jwt from 'jsonwebtoken';
+
+import db from '../db/connectDb';
+import {ResponseType} from '../models/response';
+
+import {StockListService} from './stockListService';
+
+const stockListService = new StockListService();
+
+function isMoneyNumberString(value: number): boolean {
+  return /^-?\d+(\.\d{1,2})?$/.test(value.toString());
+}
+
+
+export class PortfolioService {
+  async createPortfolio(user_id: number, name: string): Promise<ResponseType> {
+    try {
+      if (!user_id || !name) {
+        return {error: {status: 400, message: 'Name is required.'}};
+      }
+      const dup = await db.query(
+          'SELECT * FROM Portfolio WHERE user_id = $1 AND name = $2',
+          [user_id, name]);
+      if (dup.rowCount != 0) {
+        return {
+          error: {status: 400, message: 'Portfolio title already exists'}
+        };
+      }
+
+
+      let {data, error} = await stockListService.createStockList(user_id, name);
+      let port_name = name;
+      while (error) {
```

If we just check while(error) there can be an infinite loop (eg createStockList can return (some error unrelated to unique name) but the while loop doesnt know that and continues trying to change the name). This would be a problem especially if we add more errors to createStockList later on.

In fact i'd rather remove the while loop altogether and straight up let the error propagate. It's a bit better for UX if we let the user know that the name already exists and ask them to choose another name instead of us having to manually rename.
```ts
if (error) return { error };
```
Though if we absolutely had to rename it on our end, this could be done more easily by using `Date.now()` to get the unique current time and append to name.